creusot-contracts/src/std/num.rs

```diff
@@ -0,0 +1,85 @@
+use crate as creusot_contracts;
+use crate::{logic::OrdLogic, Int, Model};
+use creusot_contracts_proc::*;
+
+/// Adds specifications for checked, wrapping, saturating, and overflowing operations on the given
+/// integer type
+macro_rules! spec_type {
+    ($type:ty) => {
+        spec_op!($type, +, checked_add, wrapping_add, saturating_add, overflowing_add);
+        spec_op!($type, -, checked_sub, wrapping_sub, saturating_sub, overflowing_sub);
+        spec_op!($type, *, checked_mul, wrapping_mul, saturating_mul, overflowing_mul);
+        spec_op!($type, /, checked_div, wrapping_div, saturating_div, overflowing_div);
+    };
+}
+
+/// Adds specifications for checked, wrapping, saturating, and overflowing versions of the given
+/// operation on the given type
+macro_rules! spec_op {
+    (
+        $type:ty,
+        $op:tt,
+        $checked:ident,
+        $wrapping:ident,
+        $saturating:ident,
+        $overflowing:ident $(,)?
+    ) => {
+        extern_spec! {
+            impl $type {
+                // Checked: performs the operation on `Int`, returns `Some` if the result is between
+                // `$type::MIN` and `$type::MAX`, or `None` if the result cannot be represented by
+                // `$type`
+                #[allow(dead_code)]
+                #[ensures(
+                    (result == None)
+                    == ((@self $op @rhs) < @$type::MIN || (@self $op @rhs) > @$type::MAX)
+                )]
+                #[ensures(forall<r: $type> result == Some(r) ==> @r == (@self $op @rhs))]
+                fn $checked(self, rhs: $type) -> Option<$type>;
+
+                // Wrapping: performs the operation on `Int` and converts back to `$type`
+                #[allow(dead_code)]
+                #[ensures(
+                    @result == (@self $op @rhs).rem_euclid(2.pow(@$type::BITS)) + @$type::MIN
```

Comment:
I wonder how usable this specification will be in practice. Perhaps there's a good way to lower this to SMT so that they will be able to reason (bit vectors?)

Reply:
Probably we should have a post-condition saying its in bounds, and another saying they have the same value modulo  2^n, i.e., their difference is a multiple.

We should also have a post-condition saying that if self is in bounds, then the result is equal.